enhancements/two-nodes-openshift/2no.md

```diff
@@ -0,0 +1,398 @@
+
+#### Day 2 Proceedures
+
+As per a standard 3-node control-plane, OpenShift upgrades and `MachineConfig` changes can not be applied when the cluster is in a degraded state.
+Such operations will only proceed when both peers are online and healthy.
+
+The experience of managing a 2-node control-plane should be largely indistinguishable from that of a 3-node one.
+The primary exception is (re)booting one of the peers while the other is offline, and expected to remain so.
+
+As in a 3-node control-plane cluster, starting only one node is not expected to result in a functioning cluster.
+Should the admin wish for the control-plane to start, the admin will need to execute a supplied confirmation command on the active cluster node. 
+This command will grant quorum to the RHEL-HA components, authorizing it to fence it's peer and start etcd in as a cluster-of-one read/write mode.
+Confirmation can be given at any point and optionally make use of SSH to facilitate initiation by an external script.
+
+### API Extensions
+
+There are two related but ultimately orthogonal capabilities that may require API extensions.
+
+1. Identify the cluster as having a unique topology
+2. Tell CEO when it is safe for it to disable certain membership related functionalities
+
+#### Unique Topology
+
+A mechanism is needed for components of the cluster to understand that this is a 2 node control-plane topology which may require different handling.
+We will define a new value for the `TopologyMode` enum: `DualReplicaTopologyMode`.
+
+However `TopologyMode` is not available at the point the Agent Based Installer (ABI) performs validation.
+We will therefore additionally define a new feature gate `DualReplicaTopology` that can be enabled in `install-config.yaml`, and which ABI can use to validate the proposed cluster - such as the proposed node count.
+
+#### CEO Trigger
+
+Initially the creation of an etcd cluster will be driven in the same way as other platforms.
+Once the cluster has two members, the etcd daemon will be removed from the static pod definition and recreated as a resource controlled by RHEL-HA.
+At this point, the Cluster Etcd Operator (CEO) will be made aware of this change so that some membership management functionality that is now handled by RHEL-HA can be disabled.
+This will be achieved by having the same entity that drives the configuration of RHEL-HA use the OpenShift API to update a field in the CEO's `ConfigMap` - which can only succeed if the control-plane is healthy.
+
+To enable this flow, we propose the addition of a `managedEtcdKind` field which defaults to `Cluster` but will be set to `External` during installation, and will only be respected if the `Infrastructure` CR's `TopologyMode` is `DualReplicaTopologyMode`.
+This will allow the use of a credential scoped to `ConfigMap`s in the `openshift-etcd-operator` namespace, to make the change.
```

Comment:
What entity do you expect to be making the actual change, could we scope this change even just to a specific service account? (This is possible with ValidatingAdmissionPolicy)

Reply:
It will be performed using the cli as part of a one-shot a systemd job.

Reply:
Systemd job on one of the two nodes? How will it authenticate?

Reply:
As described it sounds like this change is initiated by one of the nodes as the one-shot systemd job. If this change is being done via CLI, where is that job getting its credentials? Also, is the plan to run the same init job on both nodes, or is that job pulling something like the IPs of the CP nodes, checking if it's the first node, and then taking on these configuration changes?

How do we detect if something goes wrong here?

Reply:
It should be possible to run the same init job on each node, but I think this thread is moot if we move to an operator-like mechanism as discussed in the above thread.